File: my-app/src/main/java/com/mycompany/app/binarySearch/FindFirstAndLastPositionOfElementInSortedArray.java
package com.mycompany.app.binarySearch;

/**
 * https://leetcode.com/problems/find-first-and-last-position-of-element-in-sorted-array/
 * Given an array of integers nums sorted in ascending order, find the starting and ending position of a given target value.
 * If target is not found in the array, return [-1, -1].
 * Follow up: Could you write an algorithm with O(log n) runtime complexity?
 *
 * Example 1:
 * Input: nums = [5,7,7,8,8,10], target = 8
 * Output: [3,4]
 *
 * Example 2:
 * Input: nums = [5,7,7,8,8,10], target = 6
 * Output: [-1,-1]
 *
 * Example 3:
 * Input: nums = [], target = 0
 * Output: [-1,-1]
 *
 * Constraints:
 * 0 <= nums.length <= 10^5
 * -10^9 <= nums[i] <= 10^9
 * nums is a non-decreasing array.
 * -10^9 <= target <= 10^9
 */

//典型的BinarySearch找最左，或者最右的问题
public class FindFirstAndLastPositionOfElementInSortedArray {
    public int[] searchRange(int[] nums, int target) {
        int[] res = new int[2];
        int left = search(nums, target, 0, nums.length-1, true);
        int right = search(nums, target, 0, nums.length-1, false);
        res[0] = left;
        res[1] = right;
        return res;
    }

    private int search(int[] nums, int target, int start, int end, boolean findLeft){
        //termination
        if(start>end){
            return -1;
        }

        int mid = (start + end) / 2;
        if(nums[mid]<target){
            return search(nums, target, mid+1, end, findLeft);
        }
        else if(nums[mid]>target){
            return search(nums, target, start, mid-1, findLeft);
        }
        else{
            int partialRes = -1;
            if(findLeft){
                partialRes = search(nums, target, start, mid-1, findLeft);
            }
            else{
                partialRes = search(nums, target, mid+1, end, findLeft);
            }
            if(partialRes!=-1){
                return partialRes;
            }
            else{
                return mid;
            }
        }
    }
}

File: python/sort_and_search/search/binary_search/find_first_and_last_position_of_element_in_sorted_array.py
"""
https://leetcode.com/problems/find-first-and-last-position-of-element-in-sorted-array/
Given an array of integers nums sorted in ascending order, find the starting and ending position of a given target value.
If target is not found in the array, return [-1, -1].
Follow up: Could you write an algorithm with O(log n) runtime complexity?

Example 1:
Input: nums = [5,7,7,8,8,10], target = 8
Output: [3,4]

Example 2:
Input: nums = [5,7,7,8,8,10], target = 6
Output: [-1,-1]

Example 3:
Input: nums = [], target = 0
Output: [-1,-1]

Constraints:
0 <= nums.length <= 10^5
-10^9 <= nums[i] <= 10^9
nums is a non-decreasing array.
-10^9 <= target <= 10^9
"""
#典型的BinarySearch找最左，或者最右的问题
from typing import List
class Solution:
    def _search(self, nums, target, start, end, find_left):
        l, r = start, end
        pos = -1
        while l<=r:
            mid = (l+r)//2
            if nums[mid]<target:
                l = mid + 1
            elif nums[mid]>target:
                r = mid - 1
            else:
                pos = mid 
                if find_left:
                    r = mid - 1
                else:
                    l = mid + 1
        return pos

    def search_range(self, nums: List[int], target: int) -> List[int]:
        left = self._search(nums, target, 0, len(nums)-1, True)
        right = self._search(nums, target, 0, len(nums)-1, False)
        return [left, right]

if __name__ == "__main__":
    sol = Solution()
    print(sol.search_range([5,7,7,8,8,10], 8))
    print(sol.search_range([5,7,7,8,8,10], 6))
    print(sol.search_range([5,7,7,8,8,10], 4))
    print(sol.search_range([5,7,7,8,8,10], 14))
    print(sol.search_range([], 0))

